pkg/volume/util/operationexecutor/operation_executor.go

```diff
@@ -793,15 +826,95 @@ func (oe *operationExecutor) generateUnmountDeviceFunc(
+	nodeName string,
+	actualStateOfWorld ActualStateOfWorldAttacherUpdater) (func() error, error) {
+	return func() error {
+		if !volumeToMount.PluginIsAttachable {
+			// If the volume does not implement the attacher interface, it is
+			// assumed to be attached and the the actual state of the world is
+			// updated accordingly.
+			addVolumeNodeErr := actualStateOfWorld.MarkVolumeAsAttached(
+				volumeToMount.VolumeSpec, nodeName, volumeToMount.DevicePath)
+			if addVolumeNodeErr != nil {
+				// On failure, return error. Caller will log and retry.
+				return fmt.Errorf(
+					"VerifyControllerAttachedVolume.MarkVolumeAsAttached failed for volume %q (spec.Name: %q) pod %q (UID: %q) with: %v.",
+					volumeToMount.VolumeName,
+					volumeToMount.VolumeSpec.Name(),
+					volumeToMount.PodName,
+					volumeToMount.Pod.UID,
+					addVolumeNodeErr)
+			}
+
+			return nil
+		}
+
+		// Fetch current node object
+		node, fetchErr := oe.kubeClient.Core().Nodes().Get(nodeName)
+		if fetchErr != nil {
+			// On failure, return error. Caller will log and retry.
+			return fmt.Errorf(
+				"VerifyControllerAttachedVolume failed fetching node from API server. Volume %q (spec.Name: %q) pod %q (UID: %q). Error: %v.",
+				volumeToMount.VolumeName,
+				volumeToMount.VolumeSpec.Name(),
+				volumeToMount.PodName,
+				volumeToMount.Pod.UID,
+				fetchErr)
+		}
+
+		if node == nil {
+			// On failure, return error. Caller will log and retry.
+			return fmt.Errorf(
+				"VerifyControllerAttachedVolume failed. Volume %q (spec.Name: %q) pod %q (UID: %q). Error: node object retrieved from API server is nil.",
+				volumeToMount.VolumeName,
+				volumeToMount.VolumeSpec.Name(),
+				volumeToMount.PodName,
+				volumeToMount.Pod.UID)
+		}
+
+		for _, attachedVolume := range node.Status.VolumesAttached {
+			if attachedVolume.Name == volumeToMount.VolumeName {
+				addVolumeNodeErr := actualStateOfWorld.MarkVolumeAsAttached(
+					volumeToMount.VolumeSpec, nodeName, volumeToMount.DevicePath)
+				glog.Infof("Controller successfully attached volume %q (spec.Name: %q) pod %q (UID: %q)",
+					volumeToMount.VolumeName,
+					volumeToMount.VolumeSpec.Name(),
+					volumeToMount.PodName,
+					volumeToMount.Pod.UID)
+
+				if addVolumeNodeErr != nil {
+					// On failure, return error. Caller will log and retry.
+					return fmt.Errorf(
+						"VerifyControllerAttachedVolume.MarkVolumeAsAttached failed for volume %q (spec.Name: %q) pod %q (UID: %q) with: %v.",
+						volumeToMount.VolumeName,
+						volumeToMount.VolumeSpec.Name(),
+						volumeToMount.PodName,
+						volumeToMount.Pod.UID,
+						addVolumeNodeErr)
+				}
+				return nil
+			}
+		}
+
+		// Volume not attached, return error. Caller will log and retry.
+		return fmt.Errorf("Volume %q (spec.Name: %q) pod %q (UID: %q) is not yet attached according to node status.",
+			volumeToMount.VolumeName,
+			volumeToMount.VolumeSpec.Name(),
+			volumeToMount.PodName,
+			volumeToMount.Pod.UID)
+	}, nil
```

Comment:
In this case yes. If you look at the other `generate*` methods you'll see that this permits pre-processing before the method is created and an error to be returned in that case.